### PR TITLE
Show block anchor attribute in Block Navigation

### LINKED ISSUE #16307: Show block ID attribute on navigator
If a block has an ID / anchor set, it could be displayed in the navigator to make it easier to distinguish among blocks and to reference them:

image

### PR DESCRIPTION
**Pull request by pierlon**
## Description
Added showing the block anchor ID in the block navigation.

Long anchor IDs is a potential issue that's up for discussion. At the moment I've hidden the overflow.

Fixes #16307.

## How has this been tested?
Developed and tested within the provided dockerized Wordpress.

## Screenshots 
image

## Checklist:
- [ ] My code is tested.
- [x] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [ ] My code has proper inline documentation. 
- [ ] I've included developer documentation if appropriate.

### CODE REVIEW

**Review by senadir — APPROVED** [review]
thank you for your PR @pierlon 
I like how you followed the current naming scheme that was introduced in #14420
I tested and it works great.
image

I'm approving this but since it's a design change as well, I will be waiting for another approve from someone in the design team.

File: packages/block-editor/src/components/block-navigation/index.js [hunk 1]
```diff
@@ -46,6 +46,7 @@ function BlockNavigationList( {
 							>
 								<BlockIcon icon={ blockType.icon } showColors />
 								{ blockType.title }
+								{ block.attributes.anchor && <span className="block-editor-block-navigation__item-anchor-name">{ block.attributes.anchor }</span> }
```
Comment (youknowriad) [inline packages/block-editor/src/components/block-navigation/index.js]: What if a block uses the "anchor" attribute for something else? I wonder if we need a dedicated block API to show the "caption" of a block based on its attributes instead?

Thoughts @mtias @gziolo @mcsf

**Review by youknowriad — CHANGES_REQUESTED** [review]

File: packages/block-editor/src/components/block-navigation/index.js (hunk 1 above)
Comment (gziolo) [reply packages/block-editor/src/components/block-navigation/index.js]: I confirm that this implementation assumes a bit too much about the name of attribute. Enforcing that this attribute can’t be used for anything else might be too cumbersome. We should at least ensure that anchor is one the flag in the supports object defined for the block.


By the way, I know that the mobile part of project experiments with an API which creates a version of the block description which is used for accessibility purposes. It clearly shows that title and description are too generic when we have multiple instances of the same block in one document. We should seek to find a way to make it possible to offer more flexibility on how blocks can be described/announced depending on the context.

**Comment by karmatosed** [comment]
Thanks for this, great to see it move into a PR.

I think having it change color as in original screenshot by @mtias would be great:

https://user-images.githubusercontent.com/548849/60195886-42577b80-983c-11e9-902d-9c12fa624b8c.png

File: packages/block-editor/src/components/block-navigation/index.js (hunk 1 above)
Comment (pierlon) [reply packages/block-editor/src/components/block-navigation/index.js]: I have refactored the code to check if the block supports the "anchor" feature.

File: packages/block-editor/src/components/block-navigation/index.js [hunk 2]
```diff
@@ -46,6 +47,7 @@ function BlockNavigationList( {
 							>
 								<BlockIcon icon={ blockType.icon } showColors />
 								{ blockType.title }
+								{ hasAnchor && <span className="block-editor-block-navigation__item-anchor-name">{ block.attributes.anchor }</span> }
```
Comment (youknowriad) [inline packages/block-editor/src/components/block-navigation/index.js]: Do we want to show the span if the anchor is empty?

**Review by youknowriad — COMMENTED** [review]
The "anchor" support is written as a hook, so technically speaking the "block-editor" package has no way to know that this hook exists or not. 

That said, I feel like there's a middle ground here between being completely generic (having a block API for this) or being very specific and that's what this PR is achieving. 

I think I wouldn't mind trying it like that, but I'd love thoughts from the team @mtias @mcsf

**Comment by gziolo** [comment]
> The "anchor" support is written as a hook, so technically speaking the "block-editor" package has no way to know that this hook exists or not.
> 
> That said, I feel like there's a middle ground here between being completely generic (having a block API for this) or being very specific and that's what this PR is achieving.

This is exactly what I shared in the 2nd part of my https://github.com/WordPress/gutenberg/pull/16898#discussion_r310450166. I didn't think about the fact that this is something that the block editor should be aware of. In the long run, we should provide some formal API which will allow rendering some sort of previews of the block depending on the context: visual, text-based, screen-reader optimized. At the same time, I wonder how friendly those anchors will be for scree-reader users?

**Comment by mcsf** [comment]
> This is exactly what I shared in the 2nd part of my [#16898 (comment)](https://github.com/WordPress/gutenberg/pull/16898#discussion_r310450166).

An API such as that devised by the mobile team could, in this case, be used directly by the anchor hook implementation, provided that BlockNavigation is extensible.

> I didn't think about the fact that this is something that the block editor should be aware of.

This is good to think about, but I wouldn't block this PR based on it. Maybe it serves as some consolation that the anchor hook lives in the `block-editor` module?

> In the long run, we should provide some formal API which will allow rendering some sort of previews of the block depending on the context: visual, text-based, screen-reader optimized. At the same time, I wonder how friendly those anchors will be for screen-reader users?

In my mind there's two ways to look at the problem: enhanced **previewing** methods, as you suggest, and better systematic **extraction** of the relevant content in a block.

- If we consider previewing, how will it solve the current problem in Block Navigation? Would we inject text-based previews for these blocks? If so, how would we reduce noise and focus on things like the anchor? or truncate the rich-text content?

- If we consider extraction, what can we learn from the mobile project?

**Comment by sarahmonster** [comment]
To evaluate this from a design perspective, it may help to take a few steps back and determine _what problem are we trying to solve?_ by adding this functionality. It's not immediately clear from #16307 what the core user problem that we're solving for is, but there may be some missing context there. 

My guess is that the purpose here is one or both of:
a. Providing a way of showing additional context to help users navigate through a potentially complex document tree. This may be better solved by considering #11010 than revealing anchor names, but that would likely depend on how useful anchor names are in context. 
b. Allowing users to more easily link to different parts of the page using anchor links. This is most likely a power user move, since not all users will be familiar with the concept and usage of anchor links, and it requires some manual work to accomplish them.

Working backwards, I looked for a way to set an anchor or an ID in the interface itself, but could only find a way of adding classes (without resorting to the HTML editor, that is.) Is the anchor or ID attribute something that users add themselves or are they added programmatically?

If the references are _user-defined_, they're a potentially useful tool in helping users navigate through a document. This certainly seems like the idea scenario, but I'm struggling to figure out how a user might be able to define an anchor like this... which likely means that most users won't know how to define an anchor, either.

If they're added _programmatically_, they may not have as much value to the majority of users, who  likely won't know what they mean or how to use them. In this case, we run the risk of adding noise to the interface without providing real value.

**Comment by hedgefield** [comment]
@sarahmonster you can add anchors under `Advanced` in the sidebar. And when entering a link, you can type the name of that anchor and it will appear in the list. Should be available on most blocks I think.

Screen Shot 2020-11-04 at 12 08 24 Screen Shot 2020-11-04 at 12 09 51

I've only ever used this to mark up headings so I could link directly to that section of a page. I've never thought about using it for adding context (as I've never written anything as complex in an editor that supported anchors) but now that you mention it, that might a use case - though it does feel a bit like an edge case yeah.

If I were to say what I'd use this for, it's to link to specific headings when sharing the URL to my page. So then the most important part for me is to be able to copy that link. The blue color suggested above would indicate that the anchor is clickable, if that then copies the permalink+anchor to my clipboard that would be useful.

**Comment by hedgefield** [comment]
@sarahmonster you can add anchors under `Advanced` in the sidebar. And when entering a link, you can type the name of that anchor and it will appear in the list. Should be available on most blocks I think.

Screen Shot 2020-11-04 at 12 08 24 Screen Shot 2020-11-04 at 12 09 51

I've only ever used this to mark up headings so I could link directly to that section of a page. I've never thought about using it for adding context (as I've never written anything as complex in an editor that supported anchors) but now that you mention it, that might a use case - though it does feel a bit like an edge case yeah.

If I were to say what I'd use this for, it's to link to specific headings when sharing the URL to my page. So then the most important part for me is to be able to copy that link. The blue color suggested above would indicate that the anchor is clickable, if that then copies the permalink+anchor to my clipboard that would be useful.

**Comment by paaljoachim** [comment]
What should we do with this PR?

**Comment by mtias** [comment]
Closing as this was implemented in #31992.

**Pull request closed without merging**